File: resources/demo/sample-solutions/two-sum.java
/**
 * Two Sum - Java Solutions
 * ========================
 * Problem: Given an array of integers and a target, return indices of two numbers
 * that add up to the target.
 */
import java.util.HashMap;
import java.util.Map;

class TwoSum {

    // ============================================================
    // Solution 1: Hash Table (Optimal)
    // Time: O(n), Space: O(n)
    // ============================================================
    public int[] twoSum(int[] nums, int target) {
        // Map from value to its index
        Map<Integer, Integer> seen = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];

            // Check if complement exists
            if (seen.containsKey(complement)) {
                return new int[]{seen.get(complement), i};
            }

            // Store current number
            seen.put(nums[i], i);
        }

        return new int[]{};  // No solution
    }

    // ============================================================
    // Solution 2: Brute Force (for comparison)
    // Time: O(n^2), Space: O(1)
    // ============================================================
    public int[] twoSumBrute(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    return new int[]{i, j};
                }
            }
        }
        return new int[]{};
    }

    // ============================================================
    // Test
    // ============================================================
    public static void main(String[] args) {
        TwoSum solution = new TwoSum();

        int[] result1 = solution.twoSum(new int[]{2, 7, 11, 15}, 9);
        assert result1[0] == 0 && result1[1] == 1;

        int[] result2 = solution.twoSum(new int[]{3, 2, 4}, 6);
        assert result2[0] == 1 && result2[1] == 2;

        System.out.println("All tests passed!");
    }
}

File: resources/demo/sample-solutions/two-sum.js
/**
 * Two Sum - JavaScript Solutions
 * ===============================
 * Problem: Given an array of integers and a target, return indices of two numbers
 * that add up to the target.
 */

// ============================================================
// Solution 1: Hash Map (Optimal)
// Time: O(n), Space: O(n)
// ============================================================
function twoSum(nums, target) {
  // Map from value to its index
  const seen = new Map();

  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];

    // Check if complement exists
    if (seen.has(complement)) {
      return [seen.get(complement), i];
    }

    // Store current number and index
    seen.set(nums[i], i);
  }

  return []; // No solution found
}

// ============================================================
// Solution 2: Brute Force (for comparison)
// Time: O(n^2), Space: O(1)
// ============================================================
function twoSumBrute(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) {
        return [i, j];
      }
    }
  }
  return [];
}

// ============================================================
// Solution 3: Using Object (alternative to Map)
// Time: O(n), Space: O(n)
// ============================================================
function twoSumObj(nums, target) {
  const seen = {};
  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];
    if (complement in seen) {
      return [seen[complement], i];
    }
    seen[nums[i]] = i;
  }
  return [];
}

// ============================================================
// Test
// ============================================================
console.assert(JSON.stringify(twoSum([2, 7, 11, 15], 9)) === JSON.stringify([0, 1]));
console.assert(JSON.stringify(twoSum([3, 2, 4], 6)) === JSON.stringify([1, 2]));
console.assert(JSON.stringify(twoSum([3, 3], 6)) === JSON.stringify([0, 1]));
console.log("All tests passed!");
